### ISSUE TITLE
Artifact As A Proto

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#909]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/909)
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 12, 2019, 15:46

## Background

Umbrella ticket (aka, epic) to capture the tasks required to implement [Artifact As A Proto](https://mail.gnome.org/archives/buildstream-list/2018-December/msg00148.html). See also the [proto and gRPC service outline](https://mail.gnome.org/archives/buildstream-list/2019-February/msg00027.html) on the list.

## Task description
* [x] 1) Create abstractions in BuildStream (without tests) - https://gitlab.com/BuildStream/buildstream/issues/908 and https://gitlab.com/BuildStream/buildstream/issues/955
    * Goals:  Have abstracted anything in `Element` which deals with the internals of an artifact into `Artifact` - result should not necessarily need additional tests (though more are welcome) since everything should already be covered
    * Results: There exists an `Artifact` class which is {1,2}:1 with `Element` for now (i.e. an `Element` _has_ 1 or 2 `Artifact` instances representing the strong and weak artifacts as appropriate.  No graph is present within `Artifact`
* [x] 2) Create abstractions in testutils - https://gitlab.com/BuildStream/buildstream/issues/945
    * Goals:  Have abstracted anything within the tests which deals with the internals of an artifact into a `TestArtifact` class.  This should not need additional tests, and should leave us with all tests still passing.
    * Results:  There exists a `TestArtifact` class in the `testutils` which allows tests to talk about the semantics of the internals of artifacts without actually caring about how they are represented on disk.  __Caveat__: If a test's purpose is to actually verify the physical layout then we'll need to keep that short-term but expect the test to be replaced during subsequent work.

**NB the above two tasks can be worked on in parallel**

* [x] 3) Review abstractions - no specific ticket, but this is considered complete and closed
    * Goals:  Ensure that the abstractions created in _1_ and _2_ are consistent and that there do not remain any un-abstracted assumptions about artifact layout in any of `Element` or the tests (modulo the above caveat).
    * Results:  Any missing abstractions are added, we're ready to proceed.
* [x] 4) Create ArtifactService protos, gRPC service - https://gitlab.com/BuildStream/buildstream/issues/965
    * Goals:  The artifact server service protos and gRPC servicer instances should be created, with a view to providing the basis on which the reworked buildstream code can be tested.
    * Results:  `bst-artifact-server` supports both the old and the new artifact services so that it can be merged short-term until cleaned up when the next job finishes.
* [x] 5) Re-work artifact and test artifact class to use the artifact proto - https://gitlab.com/BuildStream/buildstream/issues/974
    * Goals:  All existing physical artifact layout assumptions are removed, replaced with the artifact as a proto proto.  Use of `ReferenceService` is removed and replaced with use of `ArtifactService` along with commensurate support for the optionalities expressed in that.
    * Results:  The bulk of the work is now complete, we're using AaaP and issues will shake out.

~~* [ ] 6) Reference Service Excision~~
    ~~* Goals:  Remove the Reference Service from the code entirely, cease supporting it, cease walking old style refs for mtime updates in the `CASCache` etc.~~
    ~~* Results:  We no longer have the duality present in `bst-artifact-server` mentioned in 4.  __Caveat__: It may be worthwhile supporting both on the server-side for a while longer, for migration reasons - perhaps discuss at the time?~~

**Note:** the above point was edited to strike through the text as we agreed to continue to support the reference service (see comments below). 

* [ ] 7) Capabilities Service
    * Goals:  Creation of a proto and gRPC service to represent the capabilities of `bst-artifact-server`
    * Results:  We can query an artifact service to understand if it will accept build trees, what its policy is on storing failed results or log files, or whatever else we want to allow configuration of.

**NB the above is to be deferred and done separately to this undertaking (ie, when it's actually needed)** See https://gitlab.com/BuildStream/buildstream/issues/915

* [ ] 8) Verify remote execution behaviour: ensure it copes without the old reference storage service
    * Goals:  Ensure that remote execution continues to function, even without reference storage
    * Results:  We will be confident that this work has not resulted in RE failing.  This *may* go hand in hand with the work on 6, depending on whether or not they are intertwined.

**NB the above task to be done throughout the process, not just at the end**

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 12, 2019, 16:05

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @danielsilverstone-ct]](https://gitlab.com/danielsilverstone-ct) on Feb 13, 2019, 09:01

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 14, 2019, 13:09

marked this issue as related to #908

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 14, 2019, 13:12

I've added a list of functions that I expect to be abstracted into the artifacts class. I've not included push/pull as I'm unsure where these belong

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 14, 2019, 14:11

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 14, 2019, 14:24

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Feb 14, 2019, 14:29

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Feb 22, 2019, 14:57

mentioned in merge request !1175

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Feb 26, 2019, 18:29

mentioned in merge request !1187

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Mar 5, 2019, 15:25

marked this issue as related to #945

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 5, 2019, 15:33

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 13, 2019, 12:13

marked the task **1) Create abstractions in BuildStream (without tests) - https://gitlab.com/BuildStream/buildstream/issues/908** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 19, 2019, 17:41

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Mar 20, 2019, 11:03

marked this issue as related to #965

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 20, 2019, 16:41

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @raoul]](https://gitlab.com/raoul).hidalgocharman on Mar 22, 2019, 16:23

mentioned in merge request !1259

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @raoul]](https://gitlab.com/raoul).hidalgocharman on Mar 27, 2019, 14:06

mentioned in merge request !1252

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 28, 2019, 16:41

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Mar 28, 2019, 16:42

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Apr 12, 2019, 11:26

marked the task **2) Create abstractions in testutils - https://gitlab.com/BuildStream/buildstream/issues/945** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Apr 12, 2019, 11:27

marked the task **3) Review abstractions - no specific ticket, but this is considered complete and closed** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Apr 12, 2019, 11:27

marked the task **4) Create ArtifactService protos, gRPC service - https://gitlab.com/BuildStream/buildstream/issues/965** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on May 14, 2019, 12:37

marked this issue as related to #1024

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on May 15, 2019, 14:50

marked this issue as related to #1025

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on May 15, 2019, 14:58

#974 has landed, so we should look at ticking off point 5. Artifact operations have transitioned to the Artifact service, so I think it's ok to tick. Point 6 needs to be discussed more as requested by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) in terms of completely removing support for the reference service artifacts, we originally highlighted this as a discussions caveat at the time. It's worth noting that source cache uses reference here too.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 29, 2019, 09:28

As raised today in #1037, it would be nice to keep backwards compatibility with BuildStream 1 clients in BuildStream 2 artifact servers.

People want to try BuildStream 2, and it would be much less hassle if we needed to only keep one artifact server running for people who are using BuildStream 1 or testing out BuildStream 2 with only slightly modified projects.

I think it would be less overall effort to keep support for BuildStream 1 clients in the artifact server than it is for downstreams to have to setup multiple artifact servers and keep track of which artifact server supports which branch of their project (for bst1 or for testing bst2).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on May 29, 2019, 09:29

marked this issue as related to #1037

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on May 29, 2019, 09:36

That's fine with me. The reference storage service is very simple and should thus not require significant effort to keep alive. We should in that case keep it beyond the 2.0 release.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on May 29, 2019, 09:45

marked the task **5) Re-work artifact and test artifact class to use the artifact proto - https://gitlab.com/BuildStream/buildstream/issues/974** as completed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on May 29, 2019, 10:08

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @cs-shadow]](https://gitlab.com/cs-shadow) on Oct 10, 2019, 19:10

I think we should think about closing this issue now that we have been using the artifact protos for a while.

I realize that there are some outstanding issues, as mentioned in the description, but I'd propose tracking them separately to avoid this becoming the "do everything" ticket.

[[Gitlab user @juergbi]](https://gitlab.com/juergbi) [[Gitlab user @tpollard]](https://gitlab.com/tpollard) might one of you be able to summarize about what's left here?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 14, 2019, 09:43

As far as I can tell, this is complete. Task 7 is done, see #915. And remote execution is still functional with these changes. Closing this issue. Feel free to reopen if I've missed anything.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @juergbi]](https://gitlab.com/juergbi) on Oct 14, 2019, 09:43

closed
